File: python/benchmark/timing_benchmark.py
import time
import datetime
import os
import sys
import math
from typing import List, Tuple, Optional
import statistics
from logger import Logger, LoggingKeyword
from protoss_protocol import (
    Init, RspDer, Der,
    INPUT_LEN_HASH_TO_POINT,
    SESSION_KEY_LEN
)

def run_benchmark(iterations: int, run_id: int = 1, is_warmup: bool = False) -> Optional[Tuple[float, float, float]]:
    prefix = "Warmup" if is_warmup else f"Run {run_id}"
    print(f"{prefix}: Running Protoss protocol benchmark with {iterations} iterations...")

    # Configure test params
    password = "SharedPassword"
    P_i = b'\x00'
    P_j = b'\x01'

    # Initialize timing lists
    init_times = []
    rspder_times = []
    der_times = []

    # Run the test multiple times
    for i in range(iterations):
        try:
            # Time Init step
            start = time.perf_counter()
            res_init = Init(password, P_i, P_j)
            end = time.perf_counter()
            init_times.append(end - start)

            I = res_init.I
            protoss_state = res_init.protoss_state

            # Time RspDer step
            start = time.perf_counter()
            res_rspDer = RspDer(password, P_i, P_j, I)
            end = time.perf_counter()
            rspder_times.append(end - start)

            R = res_rspDer.R
            session_key_j = res_rspDer.get_session_key()

            # Time Der step
            start = time.perf_counter()
            session_key_i = Der(password, protoss_state, R)
            end = time.perf_counter()
            der_times.append(end - start)

            # Verify keys match only in the first iteration of the first run
            if i == 0 and run_id == 1 and not is_warmup:
                match = (session_key_i == session_key_j)
                if not match:
                    print("ERROR: Session keys don't match!")

        except Exception as e:
            print(f"Exception: {str(e)}")
            return None

    # Calculate average times in ms for this run
    avg_init_ms = statistics.mean(init_times) * 1000
    avg_rspder_ms = statistics.mean(rspder_times) * 1000
    avg_der_ms = statistics.mean(der_times) * 1000

    return (avg_init_ms, avg_rspder_ms, avg_der_ms)

def main():
    logger = Logger.get_instance()
    logger.log(LoggingKeyword.BENCHMARK, "See the benchmark_results/sodium folder for the info of this run.")

    # Default parameters
    iterations = 10000
    num_runs = 10

    # Parse optional CLI arguments: [iterations] [num_runs]
    if len(sys.argv) >= 2:
        iterations = int(sys.argv[1])
    if len(sys.argv) >= 3:
        num_runs = int(sys.argv[2])

    print("Protoss Protocol Timing Benchmark")
    print("=================================")

    # First run a warmup to avoid cold-start effects
    print("Performing warmup runs...")
    run_benchmark(100, is_warmup=True)

    # Run the benchmark multiple times to average out external variability
    print(f"\nRunning main benchmark ({num_runs} runs x {iterations} iterations)...")
    run_init = []
    run_rspder = []
    run_der = []
    run_total = []

    for r in range(1, num_runs + 1):
        result = run_benchmark(iterations, run_id=r, is_warmup=False)
        if result is None:
            print(f"ERROR: Run {r} failed, aborting.")
            return
        avg_init_ms, avg_rspder_ms, avg_der_ms = result
        run_init.append(avg_init_ms)
        run_rspder.append(avg_rspder_ms)
        run_der.append(avg_der_ms)
        run_total.append(avg_init_ms + avg_rspder_ms + avg_der_ms)

    # Calculate mean and standard deviation across runs
    mean_init = statistics.mean(run_init)
    mean_rspder = statistics.mean(run_rspder)
    mean_der = statistics.mean(run_der)
    mean_total = statistics.mean(run_total)

    std_init = statistics.stdev(run_init) if num_runs > 1 else 0.0
    std_rspder = statistics.stdev(run_rspder) if num_runs > 1 else 0.0
    std_der = statistics.stdev(run_der) if num_runs > 1 else 0.0
    std_total = statistics.stdev(run_total) if num_runs > 1 else 0.0

    # Format results
    results = []
    results.append(f"Benchmark Results with {iterations} iterations x {num_runs} runs")
    results.append(f"Hash Lengths: {INPUT_LEN_HASH_TO_POINT} bytes input for hash-to-point fn, {SESSION_KEY_LEN} bytes of session key")
    results.append("-------------------------")
    results.append(f"Avg. Init phase:     {mean_init:.3f} +/- {std_init:.3f} ms")
    results.append(f"Avg. RspDer phase:   {mean_rspder:.3f} +/- {std_rspder:.3f} ms")
    results.append(f"Avg. Der phase:      {mean_der:.3f} +/- {std_der:.3f} ms")
    results.append("-------------------------")
    results.append(f"Avg. Total time:     {mean_total:.3f} +/- {std_total:.3f} ms")
    results.append("\nRelative Cost:")
    results.append(f"Init phase:     {(mean_init / mean_total * 100):.1f}%")
    results.append(f"RspDer phase:   {(mean_rspder / mean_total * 100):.1f}%")
    results.append(f"Der phase:      {(mean_der / mean_total * 100):.1f}%")

    results_str = "\n".join(results)

    # Get current timestamp for the filename
    timestamp = datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    filename = f"benchmark_results_it{iterations}_{timestamp}.txt"

    # Save to file in benchmark_results folder
    logger.log_to_file(filename, results_str)
    print(f"\nBenchmark results saved to benchmark_results/sodium/{filename}")

    # Save logs before exit
    logger.save_logs()

if __name__ == "__main__":
    main()

File: libsodium-cpp/benchmark/timing_benchmark.cpp
#include <sodium.h>
#include <stdexcept>
#include <vector>
#include <string>
#include <iostream>
#include <chrono>
#include <iomanip>
#include <cmath>
#include <cstdlib>
#include <thread>
#include <sstream>
#include "logger.hpp"
#include "protoss_protocol.hpp"

static double calc_mean(const std::vector<double> &values)
{
    double sum = 0.0;
    for (double v : values)
        sum += v;
    return sum / values.size();
}

static double calc_stddev(const std::vector<double> &values)
{
    if (values.size() < 2)
        return 0.0;
    double m = calc_mean(values);
    double sum_sq = 0.0;
    for (double v : values)
    {
        double diff = v - m;
        sum_sq += diff * diff;
    }
    return std::sqrt(sum_sq / (values.size() - 1));
}

// Returns true on success, storing per-run averages in out parameters.
bool run_benchmark(int iterations, int run_id, bool is_warmup,
                   double &out_init_ms, double &out_rspder_ms, double &out_der_ms)
{
    if (is_warmup)
        std::cout << "Warmup: Running Protoss protocol benchmark with " << iterations << " iterations..." << std::endl;
    else
        std::cout << "Run " << run_id << ": Running Protoss protocol benchmark with " << iterations << " iterations..." << std::endl;

    // Configure test params
    std::string password = "SharedPassword";
    std::vector<unsigned char> P_i = {0x00};
    std::vector<unsigned char> P_j = {0x01};

    // Initialize timing variables
    auto init_time = std::chrono::duration<double>::zero();
    auto rspder_time = std::chrono::duration<double>::zero();
    auto der_time = std::chrono::duration<double>::zero();

    // Run the test multiple times to calculate average times
    for (int i = 0; i < iterations; i++)
    {
        try
        {
            // Time Init step
            auto start = std::chrono::high_resolution_clock::now();
            ReturnTypeInit res_init = Init(password, P_i, P_j);
            auto end = std::chrono::high_resolution_clock::now();
            init_time += end - start;

            std::vector<unsigned char> I = res_init.I;
            ProtossState protoss_state = res_init.protoss_state;

            // Time RspDer step
            start = std::chrono::high_resolution_clock::now();
            ReturnTypeRspDer res_rspDer = RspDer(password, P_i, P_j, I);
            end = std::chrono::high_resolution_clock::now();
            rspder_time += end - start;

            std::vector<unsigned char> R = res_rspDer.R;
            std::vector<unsigned char> session_key_j = res_rspDer.getSessionKey();

            // Time Der step
            start = std::chrono::high_resolution_clock::now();
            std::vector<unsigned char> session_key_i = Der(password, protoss_state, R);
            end = std::chrono::high_resolution_clock::now();
            der_time += end - start;

            // We only verify keys match in the first iteration of the first run
            if (i == 0 && run_id == 1 && !is_warmup)
            {
                bool match = (session_key_i == session_key_j);
                if (!match)
                {
                    std::cerr << "ERROR: Session keys don't match!" << std::endl;
                }
            }
        }
        catch (const std::exception &e)
        {
            std::cerr << "Exception: " << e.what() << std::endl;
            return false;
        }
    }

    // Calculate average times
    out_init_ms = std::chrono::duration<double, std::milli>(init_time).count() / iterations;
    out_rspder_ms = std::chrono::duration<double, std::milli>(rspder_time).count() / iterations;
    out_der_ms = std::chrono::duration<double, std::milli>(der_time).count() / iterations;

    return true;
}

int main(int argc, char *argv[])
{
    Logger::get_instance().log(LoggingKeyword::BENCHMARK, "See the benchmark_results/sodium folder for the info of this run.");
    if (sodium_init() < 0)
    {
        std::cerr << "Failed to initialize libsodium" << std::endl;
        return 1;
    }

    // Default parameters
    int iterations = 10000;
    int num_runs = 10;

    // Parse optional CLI arguments: [iterations] [num_runs]
    if (argc >= 2)
        iterations = std::atoi(argv[1]);
    if (argc >= 3)
        num_runs = std::atoi(argv[2]);

    std::cout << "Protoss Protocol Timing Benchmark" << std::endl;
    std::cout << "=================================" << std::endl;

    // First run a warmup to avoid cold-start effects
    std::cout << "Performing warmup runs..." << std::endl;
    double dummy_init, dummy_rspder, dummy_der;
    run_benchmark(100, 0, true, dummy_init, dummy_rspder, dummy_der);

    // Run the benchmark multiple times to average out external variability
    std::cout << "\nRunning main benchmark (" << num_runs << " runs x " << iterations << " iterations)..." << std::endl;
    std::vector<double> run_init, run_rspder, run_der, run_total;

    for (int r = 1; r <= num_runs; r++)
    {
        double avg_init, avg_rspder, avg_der;
        if (!run_benchmark(iterations, r, false, avg_init, avg_rspder, avg_der))
        {
            std::cerr << "ERROR: Run " << r << " failed, aborting." << std::endl;
            return 1;
        }
        run_init.push_back(avg_init);
        run_rspder.push_back(avg_rspder);
        run_der.push_back(avg_der);
        run_total.push_back(avg_init + avg_rspder + avg_der);
    }

    // Calculate mean and standard deviation across runs
    double mean_init = calc_mean(run_init);
    double mean_rspder = calc_mean(run_rspder);
    double mean_der = calc_mean(run_der);
    double mean_total = calc_mean(run_total);

    double std_init = calc_stddev(run_init);
    double std_rspder = calc_stddev(run_rspder);
    double std_der = calc_stddev(run_der);
    double std_total = calc_stddev(run_total);

    // Save results to file
    Logger &logger = Logger::get_instance();
    std::stringstream ss;
    ss << std::fixed << std::setprecision(3);
    ss << "Benchmark Results with " << iterations << " iterations x " << num_runs << " runs\n";
    ss << "Hash Lengths: " << INPUT_LEN_RISTRETTO_HASH_TO_POINT << " bytes input for Ristretto hash-to-point fn, " << SESSION_KEY_LEN << " bytes of session key\n";
    ss << "-------------------------\n";
    ss << "Avg. Init phase:     " << mean_init << " +/- " << std_init << " ms\n";
    ss << "Avg. RspDer phase:   " << mean_rspder << " +/- " << std_rspder << " ms\n";
    ss << "Avg. Der phase:      " << mean_der << " +/- " << std_der << " ms\n";
    ss << "-------------------------\n";
    ss << "Avg. Total time:     " << mean_total << " +/- " << std_total << " ms\n";
    ss << "\nRelative Cost:\n";
    ss << "Init phase:     " << (mean_init / mean_total * 100) << "%\n";
    ss << "RspDer phase:   " << (mean_rspder / mean_total * 100) << "%\n";
    ss << "Der phase:      " << (mean_der / mean_total * 100) << "%\n";

    // Get current timestamp for the filename
    auto now = std::time(nullptr);
    std::stringstream filename;
    filename << "benchmark_results_it" << iterations << "_" << std::put_time(std::localtime(&now), "%Y-%m-%d_%H-%M-%S") << ".txt";

    // Save to file in benchmark_results folder
    logger.log_to_file(filename.str(), ss.str());
    std::cout << "\nBenchmark results saved to benchmark_results/sodium/" << filename.str() << std::endl;

    system("pause");
    return 0;
}
